.github/workflows/ci.yml

```diff
@@ -49,13 +50,147 @@ jobs:
+      TAG_SUFFIX: ".${{ github.ref_type == 'tag' && github.ref_name || 'nightly' }}"
+    strategy:
+      fail-fast: false
+      matrix:
+        stack-version: ["24"]
+    steps:
+      - name: Log into Docker Hub
+        run: echo '${{ secrets.DOCKER_HUB_TOKEN }}' | docker login -u '${{ secrets.DOCKER_HUB_USERNAME }}' --password-stdin
+      - name: Publish multi-arch image index
+        run: |
+          for variant in '' '-build'; do
+            indexTag="heroku/heroku:${{ matrix.stack-version }}${variant}${TAG_SUFFIX}"
+            armTag="heroku/heroku:${{ matrix.stack-version }}${variant}_linux-arm64${TAG_SUFFIX}"
+            amdTag="heroku/heroku:${{ matrix.stack-version }}${variant}_linux-amd64${TAG_SUFFIX}"
+            docker manifest create "$indexTag" "$amdTag" "$armTag"
+            docker manifest push "$indexTag"
+          done
+
+  promote-tags:
+    if: github.ref_type == 'tag'
+    name: "Promote heroku-${{ matrix.stack-version }} tags"
+    needs:
+      - publish-images
+      - publish-indices
+    runs-on: ubuntu-24.04
+    strategy:
+      fail-fast: false
+      matrix:
+        stack-version: ["20", "22", "24"]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Log into Docker Hub
+        run: echo '${{ secrets.DOCKER_HUB_TOKEN }}' | docker login -u '${{ secrets.DOCKER_HUB_USERNAME }}' --password-stdin
+      - name: Install crane
+        uses: buildpacks/github-actions/setup-tools@v5.7.3
+      - name: Promote images to stable tag
+        run: |
+          destTags=( )
+          if (( ${{ matrix.stack-version }} >= 24 )); then
+            for variant in '' '-build'; do
+              for arch in 'amd64' 'arm64'; do
+                destTags+=("heroku/heroku:${{ matrix.stack-version }}${variant}_linux-${arch}")
+              done
+              destTags+=("heroku/heroku:${{ matrix.stack-version }}${variant}")
+            done
+          else
+            for variant in '' '-build' '-cnb' '-cnb-build'; do
+              destTags+=("heroku/heroku:${{ matrix.stack-version }}${variant}")
+            done
+          fi
+          for destTag in "${destTags[@]}"; do
+            srcTag="${destTag}.${{ github.ref_name }}"
+            crane copy "${srcTag}" "${destTag}"
+          done
+
+  publish-stack:
+    if: github.ref_type == 'tag'
+    name: "Publish heroku-${{ matrix.stack-version }} stack"
+    needs:
+      - promote-tags
+    runs-on: ubuntu-24.04
+    strategy:
+      fail-fast: false
+      matrix:
+        stack-version: ["20", "22", "24"]
+    env:
+      STACK: heroku-${{ matrix.stack-version }}
+      MANIFEST_APP_TOKEN: "${{ secrets.MANIFEST_APP_TOKEN }}"
+      MANIFEST_APP_URL: "${{ secrets.MANIFEST_APP_URL }}"
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - name: Pull base images
+        run: |
+          docker pull heroku/heroku:${{matrix.stack-version}}
+          docker pull heroku/heroku:${{matrix.stack-version}}-build
```

Comment:
I convinced myself that this issue was already in main. It's not though, because we build and store the image in the local store on main. I could do something similar here -- instead of pulling images from DockerHub, restore the cache and do a docker load (like we do for publishing images).

Reply:
Yeah I'd be fine with either restoring from the cache for this job, or else pulling the `.vNN` tags - up to you. (We don't have to worry about `.nightly` here given we don't upload those to Cheverny, so we'll always have a versioned tag - so either solution seems like it would work?)

Reply:
I've updated the workflow to generate (and cache) the runtime stack image files as part of the build job. I've moved the upload portion to it's own job that uses the cached runtime stack image files, so there's no opportunity to pull the wrong image/tag.

Reply:
Awesome, thank you!